Solidart v3: Refactor v3 core API and align ecosystem packages

medz opened the pull request:
before PR：#166, resolves #156 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * v3 public surface: ReadonlySignal, LazySignal, toReadonly(), toValueNotifier() interop; Rich Resource API with resolve/untilReady and state helpers.
  * New reactive collections: ListSignal, MapSignal, SetSignal.
  * Observer API for signal lifecycle.

* **Bug Fixes & Improvements**
  * Simplified Effect API and improved widget builders (SignalBuilder, Show); signals implement Listenable; prefer direct .value assignments.

* **Chores**
  * Version bumps across packages.

* **Tests**
  * Expanded coverage for signals, effects, resources, collections, devtools.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

medz committed:
Replace custom reactive system with alien_signals preset

medz committed:
Replace custom reactive system with alien_signals preset

medz committed:
Migrate to alien_signals for reactive system management

medz committed:
Migrate to alien_signals for reactive system management

medz committed:
Replace alien_signals imports with internal deps re-exports

medz committed:
Replace alien_signals imports with internal deps re-exports

medz committed:
Refactor Effect to extend preset.EffectNode directly

- Remove _AlienEffect wrapper class
- Simplify delayed callback logic
- Improve dependency cleanup in ReadableSignal
- Reorder imports consistently

medz committed:
Refactor Effect to extend preset.EffectNode directly

- Remove _AlienEffect wrapper class
- Simplify delayed callback logic
- Improve dependency cleanup in ReadableSignal
- Reorder imports consistently

medz committed:
Add v3 API implementation

medz committed:
Add v3 API implementation

medz committed:
Rename v3 module to next

medz committed:
Rename v3 module to next

medz committed:
WIP: Remove legacy next.dart and introduce simplified v3 API

medz committed:
WIP: Remove legacy next.dart and introduce simplified v3 API

medz committed:
legacy: Rename update method to didUpdate in internal signals

medz committed:
legacy: Rename update method to didUpdate in internal signals

medz committed:
Add Disposable interface and mixin for reactive primitives

Introduce a `Disposable` interface and `DisponsableMixin` (note: typo in
name) to provide a common disposal mechanism for `Signal`, `Computed`,
and `Effect`. This allows registering cleanup callbacks and ensures
proper resource disposal.

medz committed:
Add Disposable interface and mixin for reactive primitives

Introduce a `Disposable` interface and `DisponsableMixin` (note: typo in
name) to provide a common disposal mechanism for `Signal`, `Computed`,
and `Effect`. This allows registering cleanup callbacks and ensures
proper resource disposal.

medz committed:
Add v3.dart as a re-export entry point

medz committed:
Add v3.dart as a re-export entry point

medz committed:
Add Configuration interface and global autoDispose setting

- Add SolidartConfig with static autoDispose flag
- Make ReadonlySignal implement Configuration
- Add name and autoDispose parameters to Signal, LazySignal, and
  Computed
- Export Configuration and Disposable from the main library

medz committed:
Add Configuration interface and global autoDispose setting

- Add SolidartConfig with static autoDispose flag
- Make ReadonlySignal implement Configuration
- Add name and autoDispose parameters to Signal, LazySignal, and
  Computed
- Export Configuration and Disposable from the main library

medz committed:
Add name and autoDispose parameters to Effect constructor

The Effect class now implements Configuration and accepts optional name
and autoDispose parameters, defaulting autoDispose to the global
SolidartConfig.autoDispose value when not provided.

medz committed:
Add name and autoDispose parameters to Effect constructor

The Effect class now implements Configuration and accepts optional name
and autoDispose parameters, defaulting autoDispose to the global
SolidartConfig.autoDispose value when not provided.

medz committed:
Replace signal name with unique identifier

Introduce an `Identifier` class that provides a unique integer value
alongside an optional name. This replaces the simple `name` property on
signals, computed values, and effects to ensure each reactive primitive
has a distinct identifier for debugging and tracking.

medz committed:
Replace signal name with unique identifier

Introduce an `Identifier` class that provides a unique integer value
alongside an optional name. This replaces the simple `name` property on
signals, computed values, and effects to ensure each reactive primitive
has a distinct identifier for debugging and tracking.

medz committed:
Update import path to use solidart/v3.dart

medz committed:
Update import path to use solidart/v3.dart

medz committed:
Export SolidartConifg and remove unused exports from v3

medz committed:
Export SolidartConifg and remove unused exports from v3

medz committed:
Add advanced exports for v3 features

medz committed:
Add advanced exports for v3 features

medz committed:
Add isInitialized getter to LazySignal

Check isInitialized in value getter to throw a clearer error. The error
message now instructs to set `.value = <newValue>`.

medz committed:
Add isInitialized getter to LazySignal

Check isInitialized in value getter to throw a clearer error. The error
message now instructs to set `.value = <newValue>`.

medz committed:
Add toReadonly method to Signal class

medz committed:
Add toReadonly method to Signal class

medz committed:
Update TODO comments for naming discussions

medz committed:
Update TODO comments for naming discussions

medz committed:
Add TODO for code comments in v3.dart

medz committed:
Add TODO for code comments in v3.dart

medz committed:
Simplify LazySignal.value getter by removing braces

medz committed:
Simplify LazySignal.value getter by removing braces

medz committed:
Add dispose method to Effect class

Ensure Effect is properly stopped when disposed by calling preset.stop
before super.dispose.

medz committed:
Add dispose method to Effect class

Ensure Effect is properly stopped when disposed by calling preset.stop
before super.dispose.

medz committed:
Enable auto-dispose by default and implement cascade disposal

The default value of `SolidartConifg.autoDispose` is changed from
`false` to `true`. New static methods `Disposable.canAutoDispose`,
`unlinkDeps`, and `unlinkSubs` manage dependency cleanup. Signal,
Computed, and Effect dispose methods now cascade disposal to dependent
nodes when auto-dispose is enabled.

medz committed:
Enable auto-dispose by default and implement cascade disposal

The default value of `SolidartConifg.autoDispose` is changed from
`false` to `true`. New static methods `Disposable.canAutoDispose`,
`unlinkDeps`, and `unlinkSubs` manage dependency cleanup. Signal,
Computed, and Effect dispose methods now cascade disposal to dependent
nodes when auto-dispose is enabled.

medz committed:
Fix typo in DisposableMixin and SolidartConfig class names

medz committed:
Fix typo in DisposableMixin and SolidartConfig class names

medz committed:
Fix typo in SolidartConfig class name

medz committed:
Fix typo in SolidartConfig class name

medz committed:
Add untrackedValue getter to ReadonlySignal

Adds an `untrackedValue` getter to the `ReadonlySignal` interface and
implements it in `Signal` and `Computed`. For `Signal`, it returns the
current value without tracking. For `Computed`, it returns the cached
value if available, otherwise it computes the value without tracking.

medz committed:
Add untrackedValue getter to ReadonlySignal

Adds an `untrackedValue` getter to the `ReadonlySignal` interface and
implements it in `Signal` and `Computed`. For `Signal`, it returns the
current value without tracking. For `Computed`, it returns the cached
value if available, otherwise it computes the value without tracking.

medz committed:
Add disposal assertions to Signal and Computed value getters

Signal and Computed now assert they are not disposed when accessing
their value property, preventing use-after-dispose bugs.

medz committed:
Add disposal assertions to Signal and Computed value getters

Signal and Computed now assert they are not disposed when accessing
their value property, preventing use-after-dispose bugs.

medz committed:
Add custom equality comparator to signals and computed

Introduce `ValueComparator<T>` typedef and `SignalConfiguration<T>`
interface. Add `equals` parameter to `Signal`, `LazySignal`, and
`Computed` constructors, defaulting to `identical`. Update `didUpdate`
methods to use the comparator for equality checks before triggering
updates.

medz committed:
Add custom equality comparator to signals and computed

Introduce `ValueComparator<T>` typedef and `SignalConfiguration<T>`
interface. Add `equals` parameter to `Signal`, `LazySignal`, and
`Computed` constructors, defaulting to `identical`. Update `didUpdate`
methods to use the comparator for equality checks before triggering
updates.

medz committed:
Add tests for custom equals comparators in Signal and Computed

medz committed:
Add tests for custom equals comparators in Signal and Computed

medz committed:
Change default autoDispose to false and add detach flag

- Set SolidartConfig.autoDispose default to false
- Add SolidartConfig.detachEffects config flag
- Add detach parameter to Effect constructor
- Conditionally link effects in parent chain based on detach flag

medz committed:
Change default autoDispose to false and add detach flag

- Set SolidartConfig.autoDispose default to false
- Add SolidartConfig.detachEffects config flag
- Add detach parameter to Effect constructor
- Conditionally link effects in parent chain based on detach flag

medz commented:
- [ ] https://github.com/nank1ro/solidart/issues/168

medz committed:
Add Effect.delay factory and refactor Effect.run

The Effect.delay factory creates an effect that runs its callback after
a specified duration, with optional eager initialization. The Effect.run
method is extracted from the constructor to support manual execution.

medz committed:
Add Effect.delay factory and refactor Effect.run

The Effect.delay factory creates an effect that runs its callback after
a specified duration, with optional eager initialization. The Effect.run
method is extracted from the constructor to support manual execution.

medz committed:
Fix missing semicolon in Effect timer callback

medz committed:
Fix missing semicolon in Effect timer callback

medz committed:
Add fake_async dev dependency and test Effect.delay

medz committed:
Add fake_async dev dependency and test Effect.delay

medz committed:
WIP: *

medz committed:
WIP: *

nank1ro commented:
@medz let me know when the PR is ready to be reviewed 🙏💙

medz commented:
@nank1ro It's still in draft form. I'll update you when it's ready.

--- PS: I got so frustrated yesterday with the OpenAI GPT-5.1-Codex model that I didn't write any code all day to calm myself down 😂

medz committed:
Remove delayed effect support from Effect

colse #168

medz committed:
Remove delayed effect support from Effect

colse #168

medz commented:
@nank1ro I apologize for the delay. I've been quite busy with work lately and may have to wait a while to finish this PR. The core functionality is basically complete. Only some Solidart-specific features and devtools are still incomplete. Thank you for your patience.

nank1ro commented:
@medz no worries, take all the time you need 💙
i'm overworked of work like you in this last quarter 😅

Feel free to ping me for any issue related to specific features of Solidart

medz committed:
Add previous value tracking to signals and computeds

- Add `trackPreviousValue` static config and instance option
- Add `previousValue` and `untrackedPreviousValue` getters
- Update `didUpdate` and compute logic to store previous values
- Add comprehensive tests for new functionality

medz committed:
Add previous value tracking to signals and computeds

- Add `trackPreviousValue` static config and instance option
- Add `previousValue` and `untrackedPreviousValue` getters
- Update `didUpdate` and compute logic to store previous values
- Add comprehensive tests for new functionality

medz committed:
Add batch and untracked utility functions

medz committed:
Add batch and untracked utility functions

medz committed:
Add reactive collections and resource API

- Add ReactiveList, ReactiveMap, ReactiveSet with mutation tracking
- Add Resource API with fetcher and stream support
- Add useRefreshing config option for resources
- Add comprehensive test coverage for new features

medz committed:
Add reactive collections and resource API

- Add ReactiveList, ReactiveMap, ReactiveSet with mutation tracking
- Add Resource API with fetcher and stream support
- Add useRefreshing config option for resources
- Add comprehensive test coverage for new features

medz committed:
Add DevTools integration and observer pattern for signals

- Add `SolidartObserver` interface and `SolidartConfig.observers` list
- Add `devToolsEnabled` config flag and `trackInDevTools` signal option
- Send DevTools events for signal creation, updates, and disposal
- Support JSON serialization of signal values for DevTools
- Add comprehensive observer notification system
- Update DevTools extension to support v3 signals alongside v2

medz committed:
Add DevTools integration and observer pattern for signals

- Add `SolidartObserver` interface and `SolidartConfig.observers` list
- Add `devToolsEnabled` config flag and `trackInDevTools` signal option
- Send DevTools events for signal creation, updates, and disposal
- Support JSON serialization of signal values for DevTools
- Add comprehensive observer notification system
- Update DevTools extension to support v3 signals alongside v2

medz committed:
Fix LazySignal value check and add no-op mutation tests

- LazySignal now returns value when pendingValue is Some<T>
- Add tests for no-op mutations in ReactiveList, ReactiveMap,
  ReactiveSet
- Add tests for LazySignal previous value tracking
- Add tests for Resource concurrency and state extensions

medz committed:
Fix LazySignal value check and add no-op mutation tests

- LazySignal now returns value when pendingValue is Some<T>
- Add tests for no-op mutations in ReactiveList, ReactiveMap,
  ReactiveSet
- Add tests for LazySignal previous value tracking
- Add tests for Resource concurrency and state extensions

medz committed:
Add documentation for v3 reactive primitives

medz committed:
Add documentation for v3 reactive primitives

medz committed:
Bump solidart version to 3.0.0-dev.1

Update the package version in pubspec.yaml to 3.0.0-dev.1 for the next development release.

medz committed:
Bump solidart version to 3.0.0-dev.1

Update the package version in pubspec.yaml to 3.0.0-dev.1 for the next development release.

medz committed:
Update to solidart v3.0.0-dev.1

- Replace custom signal classes with direct solidart exports
- Update observer API to use ReadonlySignal and identifier
- Rename DisposeEffect to Effect and hasValue to isInitialized
- Remove toggle method and updateValue in favor of direct assignment
- Improve ValueListenable extensions with better disposal handling

medz committed:
Update to solidart v3.0.0-dev.1

- Replace custom signal classes with direct solidart exports
- Update observer API to use ReadonlySignal and identifier
- Rename DisposeEffect to Effect and hasValue to isInitialized
- Remove toggle method and updateValue in favor of direct assignment
- Improve ValueListenable extensions with better disposal handling

medz committed:
Clean v2 code and align v3 entry points

medz committed:
Clean v2 code and align v3 entry points

medz committed:
Align ecosystem to solidart v3

medz committed:
Align ecosystem to solidart v3

medz committed:
Guard SignalBuilder effect after unmount

medz committed:
Guard SignalBuilder effect after unmount

medz committed:
Merge branch 'main' into refactor/3.0-dev

medz committed:
Merge branch 'main' into refactor/3.0-dev

medz committed:
Trim devtools extension to v3 signals

medz committed:
Trim devtools extension to v3 signals

medz committed:
Adapt solidart_hooks to v3 APIs

medz committed:
Adapt solidart_hooks to v3 APIs

medz committed:
Fix v3 tests and examples

medz committed:
Fix v3 tests and examples

medz committed:
Bump v3 package versions

medz committed:
Bump v3 package versions

medz commented:
@nank1ro This PR is ready for review. 🙌

medz commented:
@nank1ro Tests and checks are green:\n- flutter analyze packages\n- flutter test packages\n- flutter test packages --coverage -r github --fail-fast (overall coverage ~74.73%)

medz commented:
@coderabbitai

medz commented on packages/flutter_solidart/lib/src/widgets/signal_builder.dart:
```diff
@@ -63,45 +63,57 @@ class SignalBuilder extends StatelessWidget {
-    scheduler,
-    detach: true,
+  late final Effect _effect = Effect.manual(
+    _runEffect,
     autoDispose: false,
-    autorun: false,
+    detach: true,
   );
```
😂, it looks like you're reviewing outdated code.

medz committed:
Update docs to v3

medz committed:
Update docs to v3

medz committed:
Bump alien_signals and fix toggle_theme icons

medz committed:
Bump alien_signals and fix toggle_theme icons

medz committed:
Enable web builds for examples

medz committed:
Enable web builds for examples

medz committed:
Handle SignalBuilder build errors

medz committed:
Handle SignalBuilder build errors

medz commented on examples/toggle_theme/pubspec.yaml:
```diff
@@ -34,7 +34,7 @@ dependencies:
   flutter:
     sdk: flutter
-  flutter_solidart: ^2.0.0
+  flutter_solidart: 3.0.0-dev.2
```
This wasn't my problem; it was an issue with the example project's code being non-standard 😂. It's fixed now.

medz committed:
Fix ReactiveMap addAll no-op check

medz committed:
Fix ReactiveMap addAll no-op check

medz committed:
Clarify auto-dispose behavior

medz committed:
Clarify auto-dispose behavior

medz committed:
Improve ReactiveMap docs

medz committed:
Improve ReactiveMap docs

medz committed:
Guard devtools JSON against cycles

medz committed:
Guard devtools JSON against cycles

medz commented on packages/solidart/lib/src/solidart.dart:
```diff
@@ -0,0 +1,2078 @@
+       trackPreviousValue =
+           trackPreviousValue ?? SolidartConfig.trackPreviousValue,
+       trackInDevTools = trackInDevTools ?? SolidartConfig.devToolsEnabled,
+       identifier = ._(name),
```
This is the dot abbreviation for Dart 3.10!

medz commented on packages/solidart/lib/src/solidart.dart:
```diff
@@ -0,0 +1,2078 @@
+       trackPreviousValue =
+           trackPreviousValue ?? SolidartConfig.trackPreviousValue,
+       trackInDevTools = trackInDevTools ?? SolidartConfig.devToolsEnabled,
+       identifier = ._(name),
```
This is the dot abbreviation for Dart 3.10!

medz commented on packages/solidart/lib/src/solidart.dart:
```diff
@@ -0,0 +1,2078 @@
+    String? name,
+    bool? detach,
+  }) : autoDispose = autoDispose ?? SolidartConfig.autoDispose,
+       identifier = ._(name),
```
This is the dot abbreviation for Dart 3.10!

medz committed:
Refine docs and JSON guard

medz committed:
Refine docs and JSON guard

medz committed:
Use idiomatic Effect disposal in docs

medz committed:
Use idiomatic Effect disposal in docs

medz committed:
Use ReactiveMap _copy helper

medz committed:
Use ReactiveMap _copy helper

medz committed:
Clarify Effect disposal guidance

medz committed:
Clarify Effect disposal guidance

medz committed:
Use untracked in Computed

medz committed:
Use untracked in Computed

medz committed:
Fix collections_test lint

medz committed:
Fix collections_test lint

medz commented:
@nank1ro This PR is ready for review, and I look forward to your feedback.

Compared to the previous version, the code has been reduced by nearly half, while functionality remains almost unchanged. Complete test code has also been rewritten.

There are a few details that require your approval. Should ReadonlySignal be called ReadSignal? I struggled with this for a long time when creating the PR because Read Signal is more like an action than a state.

Also, should ReactiveList/Set/Map be called the old SignalList/Set/Map?

In addition, I removed some APIs inherited from versions 0.x and 1.0 to keep it clean.

I did achieve my goal of keeping the code to around 1000 lines in the draft, but after adding all the comments, the code size increased to around 2000 lines 😂.

nank1ro commented:
> There are a few details that require your approval. Should ReadonlySignal be called ReadSignal? I struggled with this for a long time when creating the PR because Read Signal is more like an action than a state.

What about ReadableSignal? ReadSignal is shorter, but I agree it could be misleading

> Also, should ReactiveList/Set/Map be called the old SignalList/Set/Map?

Yes, I would not want to break just for a name. `ListSignal` is better IMHO than `ReactiveList`. At the end it's both a `Signal` and a `List`.

> The coverage decreased a lot, by more than 25%

Is it due to missing tests?

medz commented:
@nank1ro The coverage decrease is because I removed the "ignore test coverage" rule you previously intentionally ignored. Now, there are no rules in the code that ignore coverage (I removed it to maintain true coverage and constantly remind ourselves that we'll need to add coverage tests for devtools, etc., in the future).

Sacrificing readability for brevity isn't worthwhile, especially in development; using readonly is more intuitive.

I think ReactiveList is indeed more conceptual and technical. ListSignal is closer to readability.

Therefore, I will submit a commit later to change ReactiveList/Set/Map to the names you suggested.

nank1ro commented:
> The coverage decrease is because I removed the "ignore test coverage" rule you previously intentionally ignored. Now, there are no rules in the code that ignore coverage (I removed it to maintain true coverage and constantly remind ourselves that we'll need to add coverage tests for devtools, etc., in the future).

Previously I ignored all the alien_signals coverage because it was already tested in your alien_signals library.

medz committed:
Rename ReactiveList, ReactiveSet, and ReactiveMap to ListSignal,
SetSignal, and MapSignal

Update all references across the codebase including documentation,
examples, and DevTools integration to use the new names. This aligns the
collection signal naming with the core Signal class naming convention.

medz committed:
Rename ReactiveList, ReactiveSet, and ReactiveMap to ListSignal,
SetSignal, and MapSignal

Update all references across the codebase including documentation,
examples, and DevTools integration to use the new names. This aligns the
collection signal naming with the core Signal class naming convention.

medz committed:
remove unused config

medz committed:
remove unused config

medz commented on packages/solidart/lib/src/solidart.dart:
```diff
@@ -0,0 +1,2103 @@
+typedef ValueGetter<T> = T Function();
+
+/// A callback that returns no value.
+typedef VoidCallback = ValueGetter<void>;
```
@coderabbitai Solidart is a pure native Dart library and does not have the ability to import dart:ui, etc.

medz committed:
Add comprehensive collection signal tests and fix formatting

medz committed:
Add comprehensive collection signal tests and fix formatting

medz committed:
Rename reactive collection hooks to use *Signal naming

medz committed:
Rename reactive collection hooks to use *Signal naming

medz committed:
Fix unused expression warnings in test files

medz committed:
Fix unused expression warnings in test files

medz committed:
Add comprehensive tests for collection signals and utilities

- Test ListSignal methods: addAll, removeAt, removeLast, clear, sort,
  removeWhere, retainWhere, cast, toString
- Test MapSignal methods: keys, isEmpty, isNotEmpty, clear, update,
  toString
- Test SetSignal methods: iterator, toSet, cast, removeWhere,
  retainWhere, toString
- Add advanced tests for DisposableMixin and unlinkDeps
- Test DevTools observer const subclasses and Computed participation
- Test ResourceState maybeWhen, maybeMap, and toString
- Test Computed untrackedValue when not initialized

medz committed:
Add comprehensive tests for collection signals and utilities

- Test ListSignal methods: addAll, removeAt, removeLast, clear, sort,
  removeWhere, retainWhere, cast, toString
- Test MapSignal methods: keys, isEmpty, isNotEmpty, clear, update,
  toString
- Test SetSignal methods: iterator, toSet, cast, removeWhere,
  retainWhere, toString
- Add advanced tests for DisposableMixin and unlinkDeps
- Test DevTools observer const subclasses and Computed participation
- Test ResourceState maybeWhen, maybeMap, and toString
- Test Computed untrackedValue when not initialized

medz committed:
Add test coverage for edge cases and improve test reliability

- Test `SetSignal.clear` triggers effects and empties set
- Test `MapSignal.cast` returns properly typed signal
- Test `Computed.untrackedValue` returns cached value when available
- Test `ResourceState.maybeWhen` and `maybeMap` fallback to orElse
- Test `ResourceState` factories can be invoked at runtime
- Store `Disposable.dispose` reference to avoid calling method on
  disposed object
- Add coverage ignore comment to ResourceState base constructor
- Use `final _` pattern for unused values in tests

medz committed:
Add test coverage for edge cases and improve test reliability

- Test `SetSignal.clear` triggers effects and empties set
- Test `MapSignal.cast` returns properly typed signal
- Test `Computed.untrackedValue` returns cached value when available
- Test `ResourceState.maybeWhen` and `maybeMap` fallback to orElse
- Test `ResourceState` factories can be invoked at runtime
- Store `Disposable.dispose` reference to avoid calling method on
  disposed object
- Add coverage ignore comment to ResourceState base constructor
- Use `final _` pattern for unused values in tests

medz committed:
Add coverage ignore comment to SolidartObserver constructor

medz committed:
Add coverage ignore comment to SolidartObserver constructor

medz committed:
Fix capitalization in LazySignal error message

medz committed:
Fix capitalization in LazySignal error message

medz commented:
@nank1ro Done! Coverage increased to 100%. All tests passed. The only issue was that data actually sent to devtools couldn't be tested with coverage ignore flags. Also, due to system limitations, abstract classes (const) couldn't be ignored even when tested.

medz commented:
In summary:

1. This PR fully and correctly implements the autoDispose feature.
2. It cleans up legacy APIs from the 0.x era.
3. It significantly reduces maintenance difficulty, and all design elements are clearly visible.

nank1ro commented on docs-v2/src/content/docs/advanced/automatic_disposal.mdx:
```diff
@@ -6,108 +6,57 @@ sidebar:
-<Aside type="tip">**TL;DR**: Always dispose Effects and Observations, the rest will be disposed automatically.</Aside>
+<Aside type="tip">
+TL;DR:
+- Dispose Effects you create in widgets.
```
```suggestion
- Dispose all Effects manually
```

nank1ro commented on docs-v2/src/content/docs/advanced/automatic_disposal.mdx:
```diff
@@ -6,108 +6,57 @@ sidebar:
+tracked dependencies are disposed (not when any single dependency is disposed).
 
-There is a single case that the automatic disposal won't cover:
+## Disposing effects in widgets
```
Effects can be created in business logic classes, it has nothing to do with widgets.
For example, I use `Notifier` classes with `disco` and create/dispose effects from the class itself.

nank1ro commented on docs-v2/src/content/docs/advanced/automatic_disposal.mdx:
```diff
@@ -6,108 +6,57 @@ sidebar:
-  // ignore if already disposed
-  if (_disposed) return;
-  ...
+class ExampleState extends State<Example> {
```
```suggestion
class _ExampleState extends State<Example> {
```

nank1ro commented on docs-v2/src/content/docs/advanced/automatic_disposal.mdx:
```diff
@@ -6,108 +6,57 @@ sidebar:
-  if (_disposed) return;
-  ...
+class ExampleState extends State<Example> {
+  late final Effect _effect;
```
```suggestion
  late final Effect effect;
```

nank1ro commented on docs-v2/src/content/docs/advanced/automatic_disposal.mdx:
```diff
@@ -6,108 +6,57 @@ sidebar:
+  @override
+  void initState() {
+    super.initState();
+    _effect = Effect(() {
```
```suggestion
    effect = Effect(() {
```

nank1ro commented on docs-v2/src/content/docs/advanced/automatic_disposal.mdx:
```diff
@@ -6,108 +6,57 @@ sidebar:
+
+  @override
+  void dispose() {
+    _effect.dispose();
```
```suggestion
    effect.dispose();
```

nank1ro commented on docs-v2/src/content/docs/advanced/list_signal.mdx:
````diff
@@ -5,17 +5,17 @@ sidebar:
 ```dart
 final items = ListSignal([1, 2]);
 
-items.observe((previousValue, value) {
````
observe was special.
It won't fire during creation, unless `fireImmediately` was true.
It is often used to show a Toast/Dialog as a side effect, only tracking the next updates and skipping the current one.

nank1ro replied:
We may still use Effect in favor of observe, but we may need a new parameter to skip the first effect reaction.
Because requiring the user to skip the first reaction is not so easy to implement (from the user perspective) because it requires a new boolean variable.

nank1ro commented on docs-v2/src/content/docs/advanced/map_signal.mdx:
````diff
@@ -1,21 +1,44 @@
+// Later
+effect.dispose();
+items.dispose();
 ```
````
I would not place this paragrah here, it's all described in the automatic_disposal docs.

nank1ro commented on docs-v2/src/content/docs/advanced/set_signal.mdx:
```diff
@@ -2,20 +2,20 @@
 description: Learn SetSignal in solidart
 sidebar:
-  order: 1
+  order: 2
```
you can keep the order as it is, it will be ordered alphabetically (for docs with the same order)

nank1ro commented on docs-v2/src/content/docs/flutter/show.mdx:
```diff
@@ -48,7 +48,7 @@ class _SampleState extends State<Sample> {
   Widget build(BuildContext context) {
     return Show(
-      when: loggedIn,
+      when: () => loggedIn.value,
```
I would keep `call()` inside signals, because it is very convenient for the Show widget.

nank1ro commented on docs-v2/src/content/docs/flutter/signal_builder.mdx:
```diff
@@ -5,16 +5,14 @@ sidebar:
 ---
 
-import { Aside } from '@astrojs/starlight/components';
+A magic widget builder that automatically rebuilds every time a signal used inside its builder changes.
```
```suggestion
A magic widget builder that automatically rebuilds every time signals used inside its builder change.
```

nank1ro commented on docs-v2/src/content/docs/flutter/signal_builder.mdx:
```diff
@@ -38,9 +36,5 @@ class _SampleCounterState extends State<SampleCounter> {
-By default, if no __active__ signals are detected inside the builder method, an assertion error will be thrown in debug mode.
-This is to help catch mistakes where you might have forgotten to use an __active__ signal inside the builder.
-
-There may be cases where you want to disable this behavior, for example if you want it to build with the latest value of a disposed signal, in that case you can use `SolidartConfig.assertSignalBuilderWithoutDependencies = false` before `runApp`.
```
Why do you removed the assert?
It was very convenient to discover signals that are incorrectly disposed (when autoDispose is true)

medz replied:
Because I fixed the manual dispose behavior, for example, if a SignalBuilder depends on a Signal, and `a.dispose()` is called, the relationship between SignalBuilder and `a` is broken. At this point, `a` cannot be set to a new value. Furthermore, since `a` and SignalBuilder are no longer dependent after dispose, if `a` continues to be consumed by other subscribers without being disposed to update its value, the SignalBuilder will not be triggered.

The situation you encountered before was simply a side effect caused by an implementation error in the automatic destruction mechanism.

medz replied:
Rewritten and added back

nank1ro commented on docs-v2/src/content/docs/learning/computed.mdx:
```diff
@@ -56,30 +56,30 @@ class User {
 
 // just update the name, the effect above doesn't run because the age has not changed
-user.updateValue((value) => value.copyWith(name: 'new-name'));
+user.value = user.value.copyWith(name: 'new-name');
```
updateValue was giving the untracked value of the signal, this change is not equal to before.

nank1ro reviewed (changes requested):
I haven't reviewed all yet, because I found some changes I would like to see updated before continuing to review.

medz commented:
Hey @nank1ro, regarding the documentation, could you please commit directly? As you've probably noticed, my English is terrible, so I'm not very good at writing docs 😂.

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/flutter/signal_builder.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/flutter/signal_builder.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz committed:
Update docs-v2/src/content/docs/advanced/automatic_disposal.mdx

Co-authored-by: Alexandru Mariuti <alex@mariuti.com>

medz commented on docs-v2/src/content/docs/advanced/automatic_disposal.mdx:
```diff
@@ -6,108 +6,57 @@ sidebar:
+- Dispose all Effects manually
+- Signals/Computeds/Resources only auto-dispose when autoDispose is enabled and nothing depends on them.
+- When SolidartConfig.autoDispose is false, you MUST manually dispose Signals, Computeds, and Resources to avoid leaks.
+</Aside>
```
@nank1ro After making the changes based on your suggestions, @coderabbitai resubmitted the review comments. Hmm...

medz replied:
@coderabbitai In version 3, automatic and manual dispose are as follows:

Automatic Dispose: You can check the test file. When node 'a' is depended on by nodes 'b' and 'c', automatic dispose is enabled in both cases. When 'b' is automatically disposed of, because 'a' is depended on by 'c', and 'c' is still active, only 'b' itself is disconnected from 'a'.

Manual Dispose: Again, using 'a' as an example of nodes depended on by 'b' and 'c', after 'b' is disposed of, 'a' and 'c' are still active, so only 'b' is disconnected. If 'a' is disposed of, and 'b' and 'c' have no other dependencies or their dependencies can be safely disposed of, then all nodes with `autoDispose` enabled will be disconnected.

nank1ro replied:
<img width="749" height="973" alt="Screenshot 2026-01-12 at 16 20 34" src="https://github.com/user-attachments/assets/127208d5-9ce1-48ab-baa9-424a8d3e15d7" />

This was correctly mentioned in the old docs.
I suggested to always dispose Effects manually, because it requires less cognitive thinking.
You can dispose a single Effect, and don't have to dispose all dependencies. So disposing a single thing is better than disposing multiple things.

nank1ro replied:
> `Effect` auto-disposes only after ALL its tracked dependencies are disposed or disconnected

If the user calls `effect.dispose` it should always dispose, even it tracked dependencies are still alive.

medz committed:
Update v3 docs and signal builder checks

medz committed:
Update v3 docs and signal builder checks

medz commented on packages/solidart/lib/src/solidart.dart:
```diff
@@ -0,0 +1,2129 @@
+    final current = untrackedValue;
+    if (current.length == newLength) return;
+    value = List<E>.of(current)..length = newLength;
+  }
```
This is a known limitation of the Dart SDK's ListMixin, and we cannot resolve it! This issue persists even when using List! This is not a problem we can fix!

medz committed:
Add call operator to Signal, ReadonlySignal, and Computed

medz committed:
Add call operator to Signal, ReadonlySignal, and Computed

medz committed:
Add SignalBuilder and Show widget tests

medz committed:
Add SignalBuilder and Show widget tests

medz committed:
Update sidebar order for collection signal documentation

medz committed:
Update sidebar order for collection signal documentation

medz committed:
Fix SignalBuilder test line length

medz committed:
Fix SignalBuilder test line length

medz committed:
Fix call operator test to use type inference

medz committed:
Fix call operator test to use type inference

medz committed:
Fix SignalBuilder doc grammar

medz committed:
Fix SignalBuilder doc grammar

medz committed:
Document effect disposal in collection examples

medz committed:
Document effect disposal in collection examples

medz commented:
@nank1ro I've made some changes, please continue to review ❤️

nank1ro commented on packages/flutter_solidart/lib/src/core/value_listenable_signal_mixin.dart:
There was a reason why there was this mixin.
Every reactive object in flutter_solidart must extend ValueNotifier/ChangeNotifier.
This is especially needed for the `solidart_hooks` package, to be able to use the `useListenable` hook.

nank1ro commented on packages/solidart_devtools_extension/lib/main.dart:
```diff
@@ -134,16 +137,19 @@ class _SignalsState extends State<Signals> {
-              signals[data['_id']] = SignalData(
+          final kind = event.extensionKind;
+          switch (kind) {
+            case 'ext.solidart.v3.signal.created':
```
why v3?
the devtools gets packed together with the solidart package when released. There is no need to support different versions

nank1ro commented on packages/solidart_hooks/example/lib/pages/use_reactive_list.dart:
rename as before

nank1ro commented on packages/solidart_hooks/example/lib/pages/use_reactive_map.dart:
rename as before

nank1ro commented on packages/solidart_hooks/example/lib/pages/use_reactive_set.dart:
rename as before

nank1ro commented on packages/solidart/lib/src/solidart.dart:
````diff
@@ -0,0 +1,2129 @@
+  /// ```
+  ///
+  /// If you want the effect to run immediately, use the [Effect] factory.
+  Effect.manual(
````
why not calling it `lazy` instead of manual?

I still find it a bit inconvenient, given that the typical usage is just "skip the current value, trigger on next reaction". I would like a boolean parameter like "skipCurrent" or "skipFirst" (don't know the best name)/

medz replied:
We discussed this in another PR before. "Lazy" doesn't make sense, and its behavior isn't lazy at all. If you want to skip the first callback run of an effect, you can't collect dependencies. If you only want to control when the effect runs for the first time, the manual is indeed a bit cumbersome. But the purpose of the manual is to provide general and high-level control.

Explicit manual control is more cost-effective than creating requirements.

nank1ro replied:
What about reintroducing observe as an extension like before? It would cover the previous usage users have, and doesn't require the user to collect a new variable (like skipped) themselves and refactor all usages of observe.
https://github.com/nank1ro/solidart/blob/40119f4fcdf1f78d356174bc7cb836633cab3cf8/packages/solidart/lib/src/core/extensions.dart#L9C1-L34C2

If you look closely at the code, it uses a normal Effect, but skips calling the listener on creation.

nank1ro commented on packages/solidart/lib/src/solidart.dart:
```diff
@@ -0,0 +1,2129 @@
+      _resolveFuture = null;
+    });
+
+    return _resolveFuture!;
```
can you clarify what are you doing here?
what is `_resolveFuture`, why is needed, and why are you forcing it to be non-nullable?

medz replied:
`_doResolve` returns a Future, where `whenComplete` will always execute `return _resolveFuture`. At the time of `return`, `_resolveFuture` will definitely exist. Adding `!` is purely for linting purposes. `_resolveFuture` is a property member with a nullable signature (which we need), but it will always be a Future at the time of `return`. Adding an extra variable to receive and return it makes no sense. Adding only one character, `!`, is more cost-effective than adding about 10 characters.

medz committed:
Rename solidart_hooks example pages

medz committed:
Rename solidart_hooks example pages

medz committed:
Wrap Computed and Resource as Listenable

medz committed:
Wrap Computed and Resource as Listenable

medz committed:
Add Listenable Signal wrapper

medz committed:
Add Listenable Signal wrapper

medz committed:
Document listenable wrappers

medz committed:
Document listenable wrappers

medz committed:
Wrap collection signals as listenable

medz committed:
Wrap collection signals as listenable

medz committed:
Align listenable signal wrappers

medz committed:
Align listenable signal wrappers

medz committed:
Restore observe extension

medz committed:
Restore observe extension

medz committed:
Drop v3 from devtools extension names

medz committed:
Drop v3 from devtools extension names

medz committed:
Rename v3 templates

medz committed:
Rename v3 templates

medz committed:
Add ObserveSignal tests

medz committed:
Add ObserveSignal tests

medz committed:
Add listenable wrapper coverage tests

medz committed:
Add listenable wrapper coverage tests

medz committed:
Refine listenable wrapper coverage test

medz committed:
Refine listenable wrapper coverage test

medz committed:
Guard null devtools signal ids

medz committed:
Guard null devtools signal ids

nank1ro commented on docs-v2/src/content/docs/advanced/list_signal.mdx:
```diff
@@ -2,20 +2,42 @@
-  print("Items changed: $previousValue -> $value");
+Effect(() {
+  print('Items: ${items.value}');
 });
```
you added back `observe`, you can revert the example

nank1ro commented on docs-v2/src/content/docs/advanced/list_signal.mdx:
```diff
@@ -2,20 +2,42 @@
+    return;
+  }
+  // side effect here
+});
```
you added back observe so this comment is outdated

nank1ro commented on docs-v2/src/content/docs/advanced/map_signal.mdx:
```diff
@@ -1,21 +1,31 @@
+
 final items = MapSignal({'a': 1, 'b': 2});
 
-items.observe((previousValue, value) {
```
observe is back, revert

nank1ro commented on docs-v2/src/content/docs/advanced/set_signal.mdx:
```diff
@@ -2,20 +2,23 @@
 
-items.observe((previousValue, value) {
-  print("Items changed: $previousValue -> $value");
+Effect(() {
```
observe is back, revert

nank1ro commented on docs-v2/src/content/docs/learning/computed.mdx:
```diff
@@ -56,30 +56,30 @@ class User {
 
 // just update the name, the effect above doesn't run because the age has not changed
-user.updateValue((value) => value.copyWith(name: 'new-name'));
+user.value = user.untrackedValue.copyWith(name: 'new-name');
```
there is no need to use untrackedValue here, we're not inside an Effect, this comment is valid for the whole file.

nank1ro commented on docs-v2/src/content/docs/learning/effects.mdx:
```diff
@@ -13,32 +13,47 @@ The effect automatically subscribes to any signal and reruns when any of them ch
 
 <Aside type="caution">An effect is useless after it is disposed, you must not use it anymore.</Aside>
+
+## Lazy effects
```
```suggestion
## Manual effects
```

nank1ro commented on docs-v2/src/content/docs/learning/signal.mdx:
````diff
@@ -51,39 +51,24 @@ To change the value, you can use:
 readOnlyCounter.value = 2;
 ```
 
-## Observe the signal
````
I would not remove this paragrah, it is still true that we have observe, and not only Effect

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| `signal()` / `ReadSignal.call()` | `signal.value` | Call operator still works; `.value` is preferred. |
+| `signal.set(value)` | `signal.value = value` | Setter only. |
+| `signal.updateValue(fn)` | `signal.value = fn(signal.untrackedValue)` | Use `untrackedValue` to match v2 semantics. |
+| `signal.observe(listener, fireImmediately)` | `Effect(() { ... })` | Use effects for reactive side effects. |
```
No longer true

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| --- | --- | --- |
+| `ReadSignal<T>` | `ReadonlySignal<T>` | Read-only interface renamed. Use `toReadonly()` to create one. |
+| `toReadSignal()` | `toReadonly()` | Read-only conversion rename. |
+| `signal()` / `ReadSignal.call()` | `signal.value` | Call operator still works; `.value` is preferred. |
```
this is not a change then

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| `signal.set(value)` | `signal.value = value` | Setter only. |
+| `signal.updateValue(fn)` | `signal.value = fn(signal.untrackedValue)` | Use `untrackedValue` to match v2 semantics. |
+| `signal.observe(listener, fireImmediately)` | `Effect(() { ... })` | Use effects for reactive side effects. |
+| `signal.until(condition, timeout)` | — | Removed. Use `Completer` + `Effect` (example below). |
```
I would not have removed `until` which is pretty useful, given that you can `await` the returned future.

medz replied:
Restored

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| v2 API | v3 API | Notes |
+| --- | --- | --- |
+| `Computed(selector, options)` | `Computed(selector, ...)` | Options are named params. |
+| `computed()` | `computed.value` | Call operator still works; `.value` is preferred. |
```
1. what is this ? lowercase? Only `Computed` should be there
2. This is not a change

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+
+| v2 API | v3 API | Notes |
+| --- | --- | --- |
+| `Effect((dispose) { ... }, onError, options)` | `Effect(() { ... }, autoDispose:, name:, detach:)` | Callback no longer receives `dispose`. |
```
This is already true for the v2

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| `Effect((dispose) { ... }, onError, options)` | `Effect(() { ... }, autoDispose:, name:, detach:)` | Callback no longer receives `dispose`. |
+| `EffectOptions(delay: ...)` | — | Delay removed; use your own debounce/timer if needed. |
+| `effect()` to dispose | `effect.dispose()` | Call operator removed. |
+| `Effect.manual` | `Effect.manual` | Still available; call `run()` to start tracking. |
```
I don't understand the sentence at all

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| v2 API | v3 API | Notes |
+| --- | --- | --- |
+| `Resource(fetcher, source: SignalBase)` | `Resource(fetcher, source: ReadonlySignal)` | `SignalBase` removed. |
+| `ResourceOptions` | named params | `name`, `autoDispose`, `equals`, `trackInDevTools`, `useRefreshing`, `debounceDelay`, `trackPreviousState`. |
```
this was already true in the v2

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| `ResourceOptions` | named params | `name`, `autoDispose`, `equals`, `trackInDevTools`, `useRefreshing`, `debounceDelay`, `trackPreviousState`. |
+| `resource.update((state) => ...)` | `resource.state = ...` | Write the new state directly. |
+| `resource.select(selector)` | `Computed(() => ...)` | Use `Computed` to derive data. |
+| `resource.untilReady()` | — | Removed. Use `Completer` + `Effect`. |
```
This was pretty useful, why remove it?

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| `resource.update((state) => ...)` | `resource.state = ...` | Write the new state directly. |
+| `resource.select(selector)` | `Computed(() => ...)` | Use `Computed` to derive data. |
+| `resource.untilReady()` | — | Removed. Use `Completer` + `Effect`. |
+| `ResourceState.on / maybeOn` | `when / maybeWhen` | `on`/`maybeOn` removed. |
```
already valid in the v2

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| `resource.select(selector)` | `Computed(() => ...)` | Use `Computed` to derive data. |
+| `resource.untilReady()` | — | Removed. Use `Completer` + `Effect`. |
+| `ResourceState.on / maybeOn` | `when / maybeWhen` | `on`/`maybeOn` removed. |
+| `ResourceState.value` / `call()` | `asReady?.value` / `when` | Use extensions and pattern matching. |
```
nothing changed here

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| `SetSignal<T>` | `SetSignal<T>` | Copy-on-write collection. |
+| `MapSignal<K,V>` | `MapSignal<K,V>` | Copy-on-write collection. |
+| `useListSignal` | `useListSignal` | Hook for list signal. |
+| `useSetSignal` | `useSetSignal` | Hook for set signal. |
```
Nothing changed from v2 to v3

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+| v2 API | v3 API | Notes |
+| --- | --- | --- |
+| `ReadSignal.toValueNotifier()` | `ReadonlySignal.toValueNotifier()` | Now provided via extension. |
+| `ValueListenable.toSignal()` | `ValueListenable.toSignal()` | Still available via extension. |
```
nothing changed for the user

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
````diff
@@ -0,0 +1,269 @@
+  final signal = Signal(0, trackInDevTools: true);
+  ```
+- **Previous value tracking is read-driven**: `previousValue` / `previousState` update only after a tracked read.
+- **Nullable previous values**: `hasPreviousValue` was removed, so `previousValue == null` can mean either “no previous value” or “previous value was null.” Use a sentinel or wrap nullable values if you need to distinguish.
````
I would have keeped `hasPreviousValue` because it simplify to distinguish

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+- **Previous value tracking is read-driven**: `previousValue` / `previousState` update only after a tracked read.
+- **Nullable previous values**: `hasPreviousValue` was removed, so `previousValue == null` can mean either “no previous value” or “previous value was null.” Use a sentinel or wrap nullable values if you need to distinguish.
+- **Reactive collections are copy-on-write**: mutating methods replace the internal collection value. Avoid mutating `untrackedValue` directly.
+- **Effect timing and errors**: `EffectOptions.delay` and `onError` were removed. Use `Timer`/debounce and `try/catch` inside the effect.
```
why removing onError?

medz replied:
Because the delay has been removed, it can now be controlled manually.

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+- **Nullable previous values**: `hasPreviousValue` was removed, so `previousValue == null` can mean either “no previous value” or “previous value was null.” Use a sentinel or wrap nullable values if you need to distinguish.
+- **Reactive collections are copy-on-write**: mutating methods replace the internal collection value. Avoid mutating `untrackedValue` directly.
+- **Effect timing and errors**: `EffectOptions.delay` and `onError` were removed. Use `Timer`/debounce and `try/catch` inside the effect.
+- **Nested effects**: v3 adds `detach` / `SolidartConfig.detachEffects` to control whether inner effects become dependencies of their parent.
```
already present in v2

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
````diff
@@ -0,0 +1,269 @@
+  final prev = counter.untrackedPreviousValue;
+  print('from $prev to $next');
+});
+```
````
observe was added again by you

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
````diff
@@ -0,0 +1,269 @@
+**After (v3):**
+```dart
+final todos = ListSignal<String>([]);
+```
````
surely docs have been written by an AI 😅

medz replied:
yes😂，i'll revert docs

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
````diff
@@ -0,0 +1,269 @@
+  counter.value++;
+  other.value++;
+});
+```
````
already present in v2

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
````diff
@@ -0,0 +1,269 @@
+    error: (e, _) => Text('Error: $e'),
+  ),
+);
+```
````
nothing has changed

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+
+- `toReadSignal`, `ReadSignal`, `SignalBase`
+- `SignalOptions`, `ResourceOptions`, `EffectOptions`
+- `updateValue`, `observe`, `until`
```
`observe` was readded, `until` should be readded

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+- `toReadSignal`, `ReadSignal`, `SignalBase`
+- `SignalOptions`, `ResourceOptions`, `EffectOptions`
+- `updateValue`, `observe`, `until`
+- `ReactiveList`, `ReactiveMap`, `ReactiveSet`
```
they were never a thing

nank1ro commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,269 @@
+- `Debouncer` and other v2-only utils
+
+If you hit a missing API, prefer the v3 primitives (`Signal`, `Computed`,
+`Effect`, `Resource`) and the new helpers (`untracked`, `batch`).
```
no new helpers, already present in v2

nank1ro commented on examples/auth_flow/pubspec.yaml:
```diff
@@ -9,10 +9,11 @@ environment:
 resolution: workspace
 
 dependencies:
+  cupertino_icons: ^1.0.2
```
why cupertino_icons is back?

medz replied:
i'll remove

nank1ro commented on examples/github_search/pubspec.yaml:
```diff
@@ -30,10 +30,11 @@ resolution: workspace
 # the latest version available on pub.dev. To see which dependencies have newer
 # versions available, run `flutter pub outdated`.
 dependencies:
+  cupertino_icons: ^1.0.2
```
why cupertino_icons is back?

nank1ro commented on examples/infinite_scroll/pubspec.yaml:
```diff
@@ -9,10 +9,11 @@ environment:
 resolution: workspace
 
 dependencies:
+  cupertino_icons: ^1.0.2
```
why cupertino_icons is back?

nank1ro commented on examples/todos/pubspec.yaml:
```diff
@@ -9,10 +9,11 @@ environment:
 resolution: workspace
 
 dependencies:
+  cupertino_icons: ^1.0.2
```
why cupertino_icons is back?

nank1ro commented on examples/toggle_theme/pubspec.yaml:
```diff
@@ -31,10 +31,11 @@ resolution: workspace
 # the latest version available on pub.dev. To see which dependencies have newer
 # versions available, run `flutter pub outdated`.
 dependencies:
+  cupertino_icons: ^1.0.2
```
why cupertino_icons is back?

nank1ro commented on packages/flutter_solidart/example/lib/pages/lazy_counter.dart:
```diff
@@ -9,7 +9,7 @@ class LazyCounterPage extends StatefulWidget {
 
 class _LazyCounterPageState extends State<LazyCounterPage> {
-  final counter = Signal<int>.lazy(name: 'lazyCounter');
+  final counter = LazySignal<int>(name: 'lazyCounter');
```
Why do we need a new class when a factory was already pretty simple and explicit?

nank1ro commented on packages/flutter_solidart/example/lib/pages/lazy_counter.dart:
```diff
@@ -18,7 +18,7 @@ class _LazyCounterPageState extends State<LazyCounterPage> {
       body: Center(
         child: SignalBuilder(
           builder: (_, _) {
-            return switch (counter.hasValue) {
```
why this API change?

nank1ro commented on packages/flutter_solidart/example/lib/pages/list_signal.dart:
```diff
@@ -15,14 +13,6 @@ class ListSignalPage extends StatefulWidget {
 class _ListSignalPageState extends State<ListSignalPage> {
   final items = ListSignal([1, 2], name: 'items');
 
-  @override
-  void initState() {
-    super.initState();
-    items.observe((previousValue, value) {
-      print("Items changed: $previousValue -> $value");
-    });
-  }
-
```
revert

nank1ro commented on packages/flutter_solidart/example/lib/pages/map_signal.dart:
```diff
@@ -17,14 +15,6 @@ class MapSignalPage extends StatefulWidget {
 class _MapSignalPageState extends State<MapSignalPage> {
   final items = MapSignal({'a': 1, 'b': 2}, name: 'items');
 
-  @override
-  void initState() {
-    super.initState();
-    items.observe((previousValue, value) {
-      print("Items changed: $previousValue -> $value");
-    });
-  }
-
```
revert

nank1ro commented on packages/flutter_solidart/example/lib/pages/set_signal.dart:
```diff
@@ -15,14 +13,6 @@ class SetSignalPage extends StatefulWidget {
 class _SetSignalPageState extends State<SetSignalPage> {
   final items = SetSignal({1, 2}, name: 'items');
 
-  @override
-  void initState() {
-    super.initState();
-    items.observe((previousValue, value) {
-      print("Items changed: $previousValue -> $value");
-    });
-  }
-
```
revert

nank1ro commented on packages/flutter_solidart/example/lib/pages/show.dart:
```diff
@@ -20,7 +20,7 @@ class _ShowPageState extends State<ShowPage> {
         actions: [
           TextButton(
             style: TextButton.styleFrom(foregroundColor: Colors.white),
-            onPressed: loggedIn.toggle,
```
do you think the toggle extension was superfluous?

medz replied:
The toggle isn't intuitive enough. Furthermore, it only involves omitting two or three extra characters, while the SolidArt code requires exporting a new type. Also, the use case for bool toggling is very low. Therefore, I don't think it's worthwhile to create a special extension for this.

nank1ro commented on packages/flutter_solidart/example/lib/main.dart:
```diff
@@ -17,24 +17,35 @@ import 'package:flutter_solidart/flutter_solidart.dart';
   }
 }
 
+Object? _safeValue(ReadonlySignal<Object?> signal) {
```
why this? 😕

nank1ro commented on packages/flutter_solidart/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+## 3.0.0-dev.1
+
+- **BREAKING**: Remove Flutter-specific core wrappers under `src/core/*`; `flutter_solidart` now re-exports `solidart/solidart.dart` directly.
+- **BREAKING**: Replace `ListSignal`/`MapSignal`/`SetSignal` with `ReactiveList`/`ReactiveMap`/`ReactiveSet`.
```
no longer true

nank1ro commented on packages/flutter_solidart/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+
+## 3.0.0-dev.1
+
+- **BREAKING**: Remove Flutter-specific core wrappers under `src/core/*`; `flutter_solidart` now re-exports `solidart/solidart.dart` directly.
```
no longer true

nank1ro commented on packages/flutter_solidart/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+- **BREAKING**: Remove Flutter-specific core wrappers under `src/core/*`; `flutter_solidart` now re-exports `solidart/solidart.dart` directly.
+- **BREAKING**: Replace `ListSignal`/`MapSignal`/`SetSignal` with `ReactiveList`/`ReactiveMap`/`ReactiveSet`.
+- **BREAKING**: Replace `toReadSignal()` with `toReadonly()` and remove legacy `SignalBase`/`ReadableSignal` surfaces.
+- **FEAT**: Add v3 conversion extensions: `ReadonlySignal.toValueNotifier()` and `ValueListenable.toSignal()`.
```
already present in v2

nank1ro commented on packages/flutter_solidart/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+- **BREAKING**: Replace `ListSignal`/`MapSignal`/`SetSignal` with `ReactiveList`/`ReactiveMap`/`ReactiveSet`.
+- **BREAKING**: Replace `toReadSignal()` with `toReadonly()` and remove legacy `SignalBase`/`ReadableSignal` surfaces.
+- **FEAT**: Add v3 conversion extensions: `ReadonlySignal.toValueNotifier()` and `ValueListenable.toSignal()`.
+- **REFACTOR**: Rewrite `SignalBuilder` and `Show` to use v3 effects + dependency tracking.
```
nothing that affects the user

medz committed:
CHANGELOG formatting and consolidate v3 dev entries

medz committed:
CHANGELOG formatting and consolidate v3 dev entries

medz committed:
Revert docs-v2 to main

medz committed:
Revert docs-v2 to main

medz committed:
Update CHANGELOG for v3.0.0-dev.0 release

medz committed:
Update CHANGELOG for v3.0.0-dev.0 release

medz committed:
Update CHANGELOG for v3.0.0-dev.0 breaking changes

medz committed:
Update CHANGELOG for v3.0.0-dev.0 breaking changes

medz committed:
Align flutter_solidart dev versions

medz committed:
Align flutter_solidart dev versions

medz committed:
format

medz committed:
format

medz committed:
Remove unused cupertino_icons dependency

medz committed:
Remove unused cupertino_icons dependency

medz committed:
Remove cupertino_icons dependency from example apps

medz committed:
Remove cupertino_icons dependency from example apps

medz committed:
Update lazy counter example to use Signal.lazy factory

- Replace deprecated LazySignal constructor with Signal.lazy factory
- Use pattern matching for cleaner lazy signal state checks
- Simplify conditional logic in button handlers

medz committed:
Update lazy counter example to use Signal.lazy factory

- Replace deprecated LazySignal constructor with Signal.lazy factory
- Use pattern matching for cleaner lazy signal state checks
- Simplify conditional logic in button handlers

medz commented on packages/flutter_solidart/CHANGELOG.md:
```diff
@@ -1,3 +1,31 @@
+## 3.0.0-dev.0 (Unreleased)
+
+### Signals
+
+- **BREAKING**: Flutter `Signal`, `Computed`, `Resource`, and collection wrappers now implement `ValueListenable` only (the `ValueNotifier` mixins are removed).
+- **REMOVED**: `ValueNotifierSignalMixin` and `ValueListenableSignalMixin` exports.
+- **BREAKING**: `Signal.toReadSignal()` is replaced by `toReadonly()` and `ReadableSignal` is removed in favor of `ReadonlySignal`.
+- **REMOVED**: `Signal.updateValue` and `ToggleBoolSignal`.
+- **ADDED**: `LazySignal` wrapper returned by `Signal.lazy`.
+
+### Resource
+
+- **BREAKING**: Resource state helpers align with v3 (`ResourceStateExtensions` with `when`/`maybeWhen`; `on`/`maybeOn` removed).
+- **ADDED**: `Resource.resolve()` is now public (via core).
+
+### Widgets
+
+- **REFACTOR**: `SignalBuilder` reworked for v3 dependency tracking.
+- **REFACTOR**: `Show` simplified to a `StatelessWidget` backed by `SignalBuilder`.
+
+### Interop
+
+- **BREAKING**: `SignalBase.toValueNotifier()` becomes `ReadonlySignal.toValueNotifier()`, and `ValueNotifier.toSignal()` becomes `ValueListenable.toSignal()`.
+
+### Dependencies
+
+- **CHORE**: Bump `solidart` to `3.0.0-dev.0`.
+
```
alien_signal dependency updates are not important

medz commented on packages/flutter_solidart/CHANGELOG.md:
```diff
@@ -1,3 +1,31 @@
+
+### Signals
+
+- **BREAKING**: Flutter `Signal`, `Computed`, `Resource`, and collection wrappers now implement `ValueListenable` only (the `ValueNotifier` mixins are removed).
+- **REMOVED**: `ValueNotifierSignalMixin` and `ValueListenableSignalMixin` exports.
+- **BREAKING**: `Signal.toReadSignal()` is replaced by `toReadonly()` and `ReadableSignal` is removed in favor of `ReadonlySignal`.
+- **REMOVED**: `Signal.updateValue` and `ToggleBoolSignal`.
+- **ADDED**: `LazySignal` wrapper returned by `Signal.lazy`.
```
The naming convention has been reverted to v2, so no further explanation is needed. Only internal modifications have been made to v3.

medz commented on packages/solidart_lint/pubspec.yaml:
```diff
@@ -27,5 +27,5 @@ dependencies:
   test: ^1.25.2
-  solidart: ^2.0.0
-  flutter_solidart: ^2.0.0
+  solidart: 3.0.0-dev.0
+  flutter_solidart: 3.0.0-dev.0
```
If the version number is not updated in the pub workspace, flutter pub get will fail.

medz committed:
Add isInitialized property to Signal base class

medz committed:
Add isInitialized property to Signal base class

medz committed:
Simplify lazy counter pattern matching

medz committed:
Simplify lazy counter pattern matching

medz committed:
sort member

medz committed:
sort member

medz committed:
Restore until helpers in v3

medz committed:
Restore until helpers in v3

medz committed:
Split solidart.dart into part files

medz committed:
Split solidart.dart into part files

medz committed:
Fix coverage ignore markers

medz committed:
Fix coverage ignore markers

medz committed:
Update migration guide for v3 API changes

The migration guide has been completely rewritten to reflect the actual
API changes between v2 and v3. Key updates include:

- Restructured content to focus on actual public API changes rather than
  hypothetical ones
- Added detailed comparison tables for core types, naming conventions,
  and API changes
- Documented behavioral differences including autoDispose default change
  and previous value tracking
- Provided specific migration examples for Signals, Effects, Resources,
  and Flutter integration
- Added quick migration checklist for practical upgrade steps

medz committed:
Update migration guide for v3 API changes

The migration guide has been completely rewritten to reflect the actual
API changes between v2 and v3. Key updates include:

- Restructured content to focus on actual public API changes rather than
  hypothetical ones
- Added detailed comparison tables for core types, naming conventions,
  and API changes
- Documented behavioral differences including autoDispose default change
  and previous value tracking
- Provided specific migration examples for Signals, Effects, Resources,
  and Flutter integration
- Added quick migration checklist for practical upgrade steps

medz committed:
Update migration guide with correct v3 counter increment example

medz committed:
Update migration guide with correct v3 counter increment example

medz commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,169 @@
+
+## 4) Flutter integration (flutter_solidart)
+
+- `ReadableSignal` removed; use `ReadonlySignal`.
+- `ValueNotifierSignalMixin` is no longer exported.
+
+Extensions changed:
+
+- v2: `SignalBase.toValueNotifier()` and `ValueNotifier.toSignal()`
+- v3: `ReadonlySignal.toValueNotifier()` and `ValueListenable.toSignal()`
+
+## 5) Hooks (solidart_hooks)
+
+- `ReadSignal` → `ReadonlySignal`
+- `equals/comparator` replaced by `equals: ValueComparator<T>`
+
+Lifecycle change:
+
+- v2: dispose on unmount only if `autoDispose` was true
+- v3: hook-created signals are **always** disposed on unmount
+
+Use `useExistingSignal(...)` to bind external signals without disposing them.
+
+## 6) Behavioral differences
+
+- `SolidartConfig.autoDispose` default changed **true → false**
+- `SolidartConfig.equals` removed
+- `previousValue` updates only after a tracked read
+- `LazySignal` throws if read before first assignment (check `isInitialized`)
+
+## 7) Quick migration checklist
+
+- [ ] Upgrade versions in all solidart ecosystem packages
+- [ ] Replace `ReadSignal`/`SignalBase` with `ReadonlySignal`
+- [ ] Replace `toReadSignal()` with `toReadonly()`
+- [ ] Remove `updateValue` usages (use `.value = ...`)
+- [ ] Update Effect usage (no `delay/onError`, dispose via `.dispose()`)
+- [ ] Update Resource state handling (`state =`, `when/maybeWhen`)
+- [ ] Update Flutter extensions (`toValueNotifier`/`toSignal`)
+- [ ] Adjust hooks disposal expectations
```
No, it's not necessary. During the early development of v3, List/Set/MapSignal was renamed to ReactiveList/Set/Map. However, it has now been renamed back to the v2 naming convention, so no migration is required.

medz commented on docs-v2/src/content/docs/migration.mdx:
```diff
@@ -0,0 +1,169 @@
+- `previousValue` updates only after a tracked read
+- `LazySignal` throws if read before first assignment (check `isInitialized`)
+
+## 7) Quick migration checklist
+
+- [ ] Upgrade versions in all solidart ecosystem packages
+- [ ] Replace `ReadSignal`/`SignalBase` with `ReadonlySignal`
+- [ ] Replace `toReadSignal()` with `toReadonly()`
+- [ ] Remove `updateValue` usages (use `.value = ...`)
+- [ ] Update Effect usage (no `delay/onError`, dispose via `.dispose()`)
+- [ ] Update Resource state handling (`state =`, `when/maybeWhen`)
+- [ ] Update Flutter extensions (`toValueNotifier`/`toSignal`)
+- [ ] Adjust hooks disposal expectations
```
No, it's not necessary. During the early development of v3, List/Set/MapSignal was renamed to ReactiveList/Set/Map. However, it has now been renamed back to the v2 naming convention, so no migration is required.

medz commented on packages/solidart/lib/src/core/signal.dart:
```diff
@@ -1,172 +1,186 @@
+  ///
+  /// This is a lazy signal: it has no value at construction time.
+  /// Reading [value] before the first assignment throws [StateError].
+  factory Signal.lazy({
+    String? name,
+    bool? autoDispose,
+    ValueComparator<T> equals,
+    bool? trackPreviousValue,
+    bool? trackInDevTools,
```
You seem unfamiliar with Dart syntax, also known as factories. Parameter signature forwarding = entity. That is, whatever the constructor type signature of `LazySignal` is, `Signal.lazy` must also be the same. `equals` either passes through or is completely omitted!

medz commented on packages/solidart/lib/src/core/signal.dart:
```diff
@@ -1,172 +1,186 @@
+    assert(!isDisposed, 'Signal is disposed');
+    return super.get().unwrap();
+  }
 
-  /// Calls a function with the current value and assigns the result as the
-  /// new value.
-  T updateValue(T Function(T value) callback) =>
-      value = callback(_untrackedValue);
+  /// Sets the current value.
+  ///
+  /// {@macro solidart.signal-equals}
+  set value(T newValue) {
+    assert(!isDisposed, 'Signal is disposed');
+    set(Some(newValue));
+  }
 
-  /// Converts this [Signal] into a [ReadableSignal]
-  /// Use this method to remove the visility to the value setter.
-  ReadableSignal<T> toReadSignal() => this;
+  @override
+  T call() => value;
 
+  // TODO(nank1ro): See ReadonlySignal TODO, If `ReadonlySignal` rename
+  // to `ReadSignal`, the `.toReadonly` method should be rename?
   @override
-  // ignore: overridden_fields
-  final _id = ReactiveName.nameFor('Signal');
+  bool didUpdate() {
+    flags = system.ReactiveFlags.mutable;
+    final current = currentValue;
+    final pending = pendingValue;
+    if (current is Some<T> &&
+        pending is Some<T> &&
+        equals(pending.value, current.value)) {
+      return false;
+    }
+
+    if (trackPreviousValue && current is Some<T>) {
+      _previousValue = current;
+    }
```
Can't you see I retracted the v3 documentation? And my previous behavior didn't make sense. I didn't write the v3 documentation! I'm bad at writing documentation! Stop reminding me about the documentation!!!

medz commented on packages/solidart/lib/src/resources/resource.dart:
```diff
@@ -0,0 +1,325 @@
+
+  bool _isStale(int requestId) => requestId != _version || isDisposed;
+
+  void _listenStream() {
+    final requestId = ++_version;
+    _streamSubscription = stream!().listen(
+      (data) {
+        if (_isStale(requestId)) return;
+        state = ResourceState<T>.ready(data);
+      },
+      onError: (Object error, StackTrace stackTrace) {
+        if (_isStale(requestId)) return;
+        state = ResourceState<T>.error(error, stackTrace: stackTrace);
+      },
+    );
+  }
```
This is still the behavior of v2, and it has a bug. It will now be handled automatically! The test files are there! Go check the tests!!!

medz replied:
In addition, _resubscribe is synchronous code! This means that listen cannot emit new events before _resubscribe is called.

medz commented on packages/solidart/lib/src/core/disposable.dart:
```diff
@@ -0,0 +1,90 @@
+    }
+  }
+
+  @mustCallSuper
+  @override
+  void onDispose(VoidCallback callback) {
+    cleanups.add(callback);
```
It doesn't matter

medz committed:
Fix lazy signal update and resource refresh

medz committed:
Fix lazy signal update and resource refresh

medz committed:
Add observe callbacks to collection signal examples

medz committed:
Add observe callbacks to collection signal examples

medz commented:
@coderabbitai Regenerate PR Summary & title

medz committed:
Translate Chinese debug messages to English in example pages

medz committed:
Translate Chinese debug messages to English in example pages

medz committed:
Simplify example signal logging

medz committed:
Simplify example signal logging

SAGARSURI commented:
@nank1ro  when is this happening? Looking forward to this migration. Thanks for this great package 🙏🏽

nank1ro commented:
> when is this happening? Looking forward to this migration. Thanks for this great package 🙏🏽

Hope very soon, I'm working on something else right now, will come back after finishing the current work
